src/library/scala/collection/mutable/ListBuffer.scala

```diff
@@ -171,15 +171,10 @@ final class ListBuffer[A]
    *  @return   this $coll.
    */
   def += (x: A): this.type = {
-    if (exported) copy()
-    if (isEmpty) {
-      last0 = new :: (x, Nil)
-      start = last0
-    } else {
-      val last1 = last0
-      last0 = new :: (x, Nil)
-      last1.tl = last0
-    }
+    ensureUnaliased()
+    val last1 = new ::[A](x, Nil)
+    if (len == 0) start = last1 else last0.tl = last1
```

Comment:
Could the cause of speedup be that the code executes the allocation `new` and the rest of the code in parallel? In other words, now that the `new` is no longer control-dependent?

Reply:
I'd rather not guess.

There is a way to systematically find out, but I'm not good enough at using the tools to do this without burning lots of hours.

## Whitebox introspection of the JIT compiler

To reduce the extraneous noise in the output, you'd first need to strip down the benchmark even further to reduce the amount of code to a bare minumum.

  - [Install](https://builds.shipilev.net/openjdk-jdk11/) or locally build OpenJDK with the `fastdebug` configuration.
  - Configure the C2 JIT compiler to dump diagnostics.  `bench/jmh:run bench.ListsBenchmark.scalaListBufferPlusEq_212  -f1 -wi 4 -i4  -psize=10 -jvm /code/jdk/build/macosx-x86_64-server-fastdebug/images/jdk/bin/java -jvmArgs -XX:+PrintIdeal -jvmArgs -XX:PrintIdealGraphLevel=2 -jvmArgs -XX:PrintIdealGraphFile=/tmp/graph.xml`
  - Inspect the compiler optimizations with the [Ideal Graph Visualizer](https://github.com/openjdk/jdk/tree/master/src/utils/IdealGraphVisualizer)
  - Compare fast and slow versions above.

## Blackbox comparison of the generated assembly

Someone used to reading x86 assembly (in particular that generated by HotSpot) could likely just glance at https://gist.github.com/retronym/3dc4d3c1918c4f4ec46f958326484fc3 and explain why the fast version is faster.